[ENH] New Dataset Downloaders for TSF and TSC Datasets

jgyasu opened the pull request:
We recently copied TSF and TSC datasets from different sources to sktime's Hugging Face repository to make downloads more reliable and reduce the payloads by not downloading datasets in redundant formats. This PR adds the utility for downloading those datasets from HF repo. All the public interfaces remain the same.

The `DataDownloader` class follows the strategy pattern and it can be instantiated with both the HF repo and the fallback URLs, it first tries to download from the HF repo and upon failure, it tries all the original sources' URLs.

jgyasu commented:
Hmm, weird, m5 datasets have two download locations, explains why the tests are failing.

jgyasu commented:
> Nice! does this replace #8404?

Yes, but also adds on to handle TSF datasets as well.

fkiraly commented on sktime/datasets/_dataset_downloader.py:
```diff
@@ -0,0 +1,248 @@
+            raise Exception(error_msg)
+
+
+class DatasetDownloader:
```
instead of a hard-coded downloader class, I would go with the composite pattern:

`FallbackDownloader(sequence=List[DatsetDownloadStrategy, retries=3)` - with `FallbackDownloader` itself inheriting from `DatasetDownloadStrategy`!

This way you keep it flexible and extensible if we later want to add more strategies to download from.

jgyasu replied:
Sorry, I do not understand. Isn't `URLDownloader` the `FallbackDownloader`? And it is inheriting from `DatasetDownloadStrategy`

jgyasu replied:
Makes sense, I am reading up some design articles, I will update the design.

fkiraly commented on sktime/datasets/_dataset_downloader.py:
```diff
@@ -0,0 +1,248 @@
+            defaults to a 'local_data' folder in the current working directory.
+        force_download : bool, default=False
+            Whether to force re-download the dataset even if it's cached locally.
+        token : str, optional
```
should the token not be part of the `__init__` params?

If we use the strategy pattern, the args of `download_dataset` should be the same across all strategies.

fkiraly reviewed (commented):
Question: I see using HF requires a token. What is the strategy for token handling, e.g., does an individual user need to set a token first?

jgyasu commented:
> I see using HF requires a token

It doesn't. Token is recommended by HF but it is not required for public repositories, they are only strict for private repositories. I don't know why they recommend it even.

jgyasu commented on sktime/datasets/_single_problem_loaders.py:
```diff
@@ -1601,10 +1598,17 @@ def load_m5(
         if not os.path.exists(os.path.join(extract_path, "m5-forecasting-accuracy")):
```
Hmm, I haven't touched the old code yet, but I will look into it.

jgyasu commented:
I see why tests are failing, will fix them soon.

jgyasu commented:
URLDownloader seem to fail because the URLs are of the form, [https://zenodo.org/record/4656589/files/m4_hourly_dataset.zip/m4_hourly_dataset.zip](https://zenodo.org/record/4656589/files/m4_hourly_dataset.zip/m4_hourly_dataset.zip) instead of [https://zenodo.org/record/4656589/files/m4_hourly_dataset.zip](https://zenodo.org/record/4656589/files/m4_hourly_dataset.zip), this error arises in the first place because of the different formatting of URLs of different sources and I am working on a fix. 

But HFDownloader should work, it is working on my local system.

jgyasu commented on sktime/datasets/_dataset_downloader.py:
```diff
@@ -0,0 +1,303 @@
+        **kwargs
+            Additional keyword arguments passed to `snapshot_download`.
+
+        Returns
```
Can we perhaps separate exceptions in the following way?
```python
        try:
            snapshot_download(
                repo_id=self.repo_name,
                repo_type=self.repo_type,
                allow_patterns=f"{dataset_name}/**",
                local_dir=download_path,
                force_download=force_download,
                token=self.token,
            )

            if not local_dataset_path.exists():
                raise ValueError(
                    f"Dataset folder '{dataset_name}' not found"
                    " in repository '{self.repo_name}'"
                )

            return local_dataset_path
            except (RepositoryNotFoundError, HfHubHTTPError) as e:
                raise RepositoryAccessError(f"Failed to access repository: {e}")
            except Exception as e:
                raise DatasetDownloadError(f"Unexpected error during download: {e}")
```

fkiraly reviewed (changes requested):
Great!

A few small remarks:

* can we rename the method `download_dataset` to `download`?
* there is some repeated boilerplate in the `download_dataset` method. I would recommend to split it in a `download` for boilerplate and `_download` for specific methods
    * the inner one does not need `force_download` if we deal with the cache check in the boilerplate.
* the HF downloader does not seem to do caching, is that correct? I think it should, e.g., `if extract_path_exists`
* can we move the imports of the HF hub into the methods? Then we do not need `_safe_import`.

For discussion: right now, users do not need `huggingface_hub` for the downloads, so if we would release this, their code would break.
A pattern that might avoid this:

* add soft dependency checks to the download strategies, using `python_dependencies`
* The `FallbackDownloader` only runs the strategies with dependencies present, but gives the user a clear warning if a strategy is skipped due to missing dependency

What do you think? This way users upgrading after the next release get the warning that they want to use the HF hub, but the code does not break.

jgyasu commented:
@fkiraly thanks for the comments, I have addressed all of them in the recent commit, please let me know if you had any more design comments.

jgyasu commented:
> > I am confused why some of these tests are failing in the CI, they seem to be passing on my local environment.
> 
> Silly question - the screenshot does not seem to show failing tests, only passing tests. What exactly is failing?

The screenshot was from local system where the tests were passing, I saw the tests fail in the GitHub testing pipeline. But it's all good now, fixed everything.

jgyasu commented:
I decided to test the strategies separately again, and HFDownloader is not working anymore, perhaps it is to do something with the fact that `_check_soft_dependencies` is not being used properly?

But this also prompts me to write tests for both the strategies separately as well instead of relying on the tests for data loaders.

fkiraly commented:
Some of the tests are failing, it looks like failures specific to the new code.

Have you investigated the failures? If so, what is the precise reason? Are there any bugs?

jgyasu commented:
> Have you investigated the failures? If so, what is the precise reason? Are there any bugs?

The tests which fail are the tests I have written to test `HuggingFaceDownloader` and `URLDownloader` separately because I figured out `HuggingFaceDownloader` wasn't working. It stopped working after I stopped importing `huggingface-hub` using`safe_import`. If I use `_safe_import` then the tests will pass again. I can't understand the reason behind this.

jgyasu commented:
fixed, it was a typo in the module name

fkiraly commented:
hm, so the soft dep tag will have no effect in the automated test system, since:

* your tests are not using the `scikit-base` machinery
* your dataset collection classes do not inherit from `BaseObject`

I would not necessarily add that though, but instead see if we can manually change the tests.

I would also advise, I might have said this already somewhere: for a user who does not have `huggingface_hub` installed, I would suggest:

* skip download options in the fallback programmatically based on presence of soft dependencies
* if one is skipped give the user a warning, that we recommend them to install softdep xyz for prefered download location abc

jgyasu commented:
> but instead see if we can manually change the tests
the tests which fail are written by me, I will remove them, I have verified the run on my local system

> I would also advise, I might have said this already somewhere: for a user who does not have huggingface_hub installed, I would suggest:
But I had already added these changes yesterday

jgyasu commented on sktime/datasets/_dataset_downloader.py:
```diff
@@ -0,0 +1,334 @@
+        errors = []
+
+        for i, strategy in enumerate(self.strategies):
+            if isinstance(strategy, HuggingFaceDownloader) and not strategy.available:
+                next_strategy = type(self.strategies[i + 1]).__name__
+                warnings.warn(
+                    f"huggingface_hub is not available, please install it to"
+                    f" use Hugging Face for dataset downloads, skipping "
+                    f"HuggingFaceDownloader, using {next_strategy}."
+                )
+                continue
```
@fkiraly FYI

fkiraly replied:
yes, I know, the key to this is "programmatically", so not making a specific assumption about another class, but using properties instead. What I would do:

* add the `python_dependencies` tag
* then use `_check_soft_dependencies` on that tag
* return an error message that works in general, e.g., `In {name of downloader}, skipping {specific strategy} as it requires {package_name}. We recommend this location as download mirror no.{i}, to use this please install dependencies {package_name}, by pip install {package_name}`.

jgyasu replied:
Got it, makes sense. Thank you!

fkiraly commented on sktime/datasets/_dataset_downloader.py:
```diff
@@ -0,0 +1,342 @@
+        download_path.mkdir(parents=True, exist_ok=True)
+        local_dataset_path = download_path / dataset_name
+
+        if local_dataset_path.exists():
```
I think this can be simplified:

```python
if not local_dataset_path.exists() or force_download:
    self._download(dataset_name, download_path, **kwargs)
return local_dataset_path
```

fkiraly reviewed (changes requested):
This is great, only very minor questions and requests:

* simplification of the `download` boilerplate logic above
* I think we should be testing the individual downloaders for the strategy pattern
    * behaviour if downloading for first time: file is not there yet; then file gets downloaded
    * behaviour if downloading for second time: file already exists; both settings of `force_download` (for `force_download=True`, also check file contents)
    * loop this over instances of all strategies
* question, not blocking: `DatasetDownloader` is now just an alias for `FallbackDownloader`. Is it still needed, or can it be completely replaced? If your intention was to have a shorthand, it is fine for me; though if you refactored and simply did not notice this, then you may consider removing.

fkiraly commented on sktime/datasets/forecasting/m5_competition.py:
```diff
@@ -93,9 +94,14 @@ def path_to_data_dir(self):
     def _download_if_needed(self):
         """Download the data if it is not already downloaded."""
         if not self.path_to_data_dir.exists():
```
is this and other `if not path.exists` not already covered by the boilerplate in `download`? Can it be removed, or does it serve another purpose?

jgyasu commented:
I have refactored the code with the suggestions, added tests as well. But I have 2 questions,

1. The tests I have written will fail for the `HuggingFaceDownloader` because `huggingface-hub` is not present in the testing pipeline, how can I fix this? Do we need to skip testing this specific strategy?

2. When the `force_download=True`, how do we know that the dataset is downloaded again and is not the existing one? I don't think there is any way to test the hash changes, or is there? I had this idea of testing by the timestamp but I don't know if it is a good practice?

fkiraly commented on .github/workflows/test_datasets.yml:
```diff
@@ -73,7 +73,7 @@ jobs:
           python-version: ${{ matrix.python-version }}
       - name: dependencies installation step
-        run: python3 -m pip install .[tests,datasets]
+        run: python3 -m pip install .[tests,datasets] huggingface-hub
```
yes, insofar it is the place I meant

fkiraly reviewed (approved):
Very nice PR! All points addressed, stringent and nice composable architecture!

Testing also seems to work!
(I assume you checked in the logs that they indeed run, right?)

fkiraly reviewed (changes requested):
Thanks, great!

Now that I understand the intended contract (thanks for explaining), I think it is inconsistent around the `dataset_name` argument of `download`.

This is inconsistently used, for instance, in `URLDownloader`, you seem to pass the `dataset_name` as part of the path, so passing it again in `download` would be unnecessary (and it is in fact ignored when retrieving the file). While in `HF_REPO_NAME` you do not pass the information to the `__init__`.

I think the deeper reason for this is that you have not thought through the implied contract in terms of where files are located at the download source, and where they are written on the hard drive.

I would suggest you add the "guarantee" of which files are created where to the docstring as well, and explain more precisely what `dataset_name` does in the `download` method.

I would simply remove it entirely, although I have not fully thought it through.

jgyasu commented:
> I would simply remove it entirely, although I have not fully thought it through.

`URLDownloader` does not use `dataset_name` for downloading datasets, it is simply a function of a list of URLs, but `HuggingFaceDownloader` uses `dataset_name` to download the specific folders from HuggingFace, so I don't think we can remove it entirely. 

> This is inconsistently used, for instance, in URLDownloader, you seem to pass the dataset_name as part of the path, so passing it again in download would be unnecessary (and it is in fact ignored when retrieving the file)

This is true, we can definitely remove `dataset_name` from the `URLDownloader` but we need to use it for `HuggingFaceDownloader`, so I was of the notion that removing it from one strategy will make the contract inconsistent, please correct me if I am wrong?

> I think the deeper reason for this is that you have not thought through the implied contract in terms of where files are located at the download source, and where they are written on the hard drive.
>...
> I would suggest you add the "guarantee" of which files are created where to the docstring as well

I have thought about this, but the existing loader functions extract datasets in a few different locations so we either change that behaviour and then add the guarantee to our downloader classes or we do not. For example, M5 datasets are downloaded and extracted at `sktime/m5-forecasting-accuracy`, while the classification and forecasting datasets are downloaded and extracted at `sktime/datasets/data/dataset-name`. 

Currently the extraction path is handled and documented by the loader functions and I think it is best if they handle it. Downloaders are just generic downloaders which accepts an extraction path from the loader function and extracts the dataset there.

>  While in HF_REPO_NAME you do not pass the information to the __init__.

I do not understand this point, I don't think we have a class or variable named `HF_REPO_NAME`. In `HuggingFaceDownloader`, `repo_name` is passed, while in `DatasetDownloader`, it is named `hf_repo_name`, this is not an error, just an oversight, so if you meant this then I will make this argument consistent.

jgyasu commented on sktime/datasets/_dataset_downloader.py:
```diff
@@ -0,0 +1,302 @@
+        for i, strategy in enumerate(self.strategies):
+            if not strategy.available:
+                strategy_name = type(strategy).__name__
+                soft_dependency = strategy.get_tag("python_dependencies")
```
@fkiraly I just noticed this, here `get_tag` won't work because strategies aren't the direct or indirect child of `BaseObject`. Will it be a good idea to make `DatasetDownloadStrategy` a child of `BaseObject`?

fkiraly replied:
why did this not throw an error previously?

Yes, I think it might make sense to inherit from `BaseObject` if you want to use tags; but also, we should add tests to cover these lines.

jgyasu replied:
It did not throw an error because the test for individual downloaders gets skipped if a soft dep is not available, see,
```python
def test_downloader_strategy_behavior(tmp_path, strategy):
    """Test different dataset download strategies."""
    if not strategy.available:
        pytest.skip(f"{type(strategy).__name__} not available")
```
If we do not skip the test then it will fail.

> we should add tests to cover these lines

Test what exactly?

fkiraly replied:
> Test what exactly?

The failure case.

You can do this by using `pytest.raises` instead of `pytest.skip`.

fkiraly reviewed (changes requested):
Two pieces of feedback, high level but imo important:

* can you please ensure that the class docstrings make clear what the class does upon `download`. Where does it look, which files does it create. Please be exact in your statements.
    * I am not giving deeper feedback at the moment since the lack of clarity in docstrings prevents me from understanding the deeper intention, e.g., the exact function of the `dataset_name` argument in `download`. 
* the tests related to the loaders seem to be failing, please check.

fkiraly reviewed (changes requested):
Thanks for the careful documentation - I now understand the intention!

Based on this, I think the design is redundant, that is, the `__init__` args, `dataset_name` and `download_path` are partially duplicative, and in particular the `dataset_name` is in too many places and redundant. I have looked into how it is used in `_single_problem_loader` and have not found a reason why we should allow this redundance in parametrization.

I would request:

* in `HuggingFaceDownloader`, add a `folder_name` to the `__init__`, this is the path inside the repo where the downloader looks. In `download`, make `download_path` the full path where the file is downloaded to, i.e., "downloaded to" would simply be `{download_path}/`. Remove `dataset_name` altogether.
* in `URLDownloader, remove `dataset_name` altogethe as well, it is not needed for usage. The full path where the file is downloaded should simply be `{download_path}/`.
* Change usage in downstream locations accordingly. Please let me know if you think that is not easily possible, or if there are any other obstructions.

fkiraly commented:
(tests are failing, I advise you try to run the downloaders locally)

jgyasu commented:
I have removed the `dataset_name` arg entirely now. Added a helper function to generate a dataset folder path on the hard drive, for `URLDownloader` it extracts the dataset folder path from the URL and for `HuggingFaceDownloader`, it gets it from `folder_name` arg which the class is initialised with.

jgyasu commented:
> Do you think this is "cleaner", or does it feel worse, in your opinion?

Hmm, I have mixed feelings. We're already passing an additional parameter (`folder_name`) for Hugging Face datasets, so we could have reused it for the URL downloader too - just rename `dataset_name` to `folder_name`.

That said, I think it's cleaner to use parameters strictly for their intended purpose. For example, `folder_name` is primarily meant for Hugging Face downloads, and creating a folder on disk is more of a secondary concern. In the case of URL downloads, we can derive the dataset name from the URL itself, so `folder_name` isn't strictly necessary there.

Though, I do not feel very confident of my design opinions yet solely due to my lesser experience in designing software.

fkiraly commented on sktime/datasets/_dataset_downloader.py:
```diff
@@ -0,0 +1,431 @@
+            If the expected dataset folder is not found in the downloaded snapshot.
+        """
+        if not self.available:
+            raise ModuleNotFoundError
```
this is an uninformative error without message - should this not be raised with an informative error message, and already at `__init__`?

jgyasu replied:
hmm, I realised there is a better way to do this, I removed this check and just turned `severity="none"` to `severity="error"` in the `__init__`

`_check_soft_dependencies("huggingface-hub", severity="error")`

jgyasu replied:
Hmm, I think it is because the error is raised at `__init__`, I tried putting a check in `__init__` too, like 
```python
if not self.available:
    raise ModuleNotFoundError("The required module is not available. Please ensure it is installed.")
```
But the tests still does not pass, however with this check in `_download`, the tests pass.

fkiraly reviewed (approved):
Absolutely great, I think this is ready now!

Just a minor detail, in the `download` docstring, it would be nice to describe side effects, e.g., which file is impacted (created or overwritten).
